File: clojure/secret-handshake/src/secret_handshake.clj
(ns secret-handshake)

(defn contains-bit? [n bit]
  (not (zero? (bit-and n (bit-shift-left 1 bit)))))

(def code-ops
  [#(conj % "wink")
   #(conj % "double blink")
   #(conj % "close your eyes")
   #(conj % "jump")
   reverse])

(defn bit-encoder [n]
  (fn encode-bit [acc [i op]]
    (if (contains-bit? n i)
      (op acc)
      acc)))

(defn commands [n]
  (->> code-ops
       (map-indexed vector)
       (reduce (bit-encoder n) [])))

File: haskell/secret-handshake/src/SecretHandshake.hs
module SecretHandshake (handshake) where

import Data.List as List

handshake :: Int -> [String]
handshake n =
  reverse $ foldl action [] $ binaryComponents n

action :: [String] -> Int -> [String]
action xs i =
  case i of
    1 -> "wink" : xs
    2 -> "double blink" : xs
    4 -> "close your eyes" : xs
    8 -> "jump" : xs
    16 -> reverse xs
    _ -> error "Found components greater than 16! Probably you entered a too large number."

binaryComponents :: Int -> [Int]
binaryComponents n =
  filter (/= 0) . List.zipWith (*) (iterate (*2) 1) $ (binary n)

binary :: Int -> [Int]
binary 0 = []
binary n =
  (rem n 2) : (binary (div n 2))
